README
NAME
    HTTP::Daemon::UNIX - HTTP::Daemon over Unix sockets

VERSION
    This document describes version 0.05 of HTTP::Daemon::UNIX (from Perl
    distribution HTTP-Daemon-UNIX), released on 2014-12-05.

SYNOPSIS
     use HTTP::Daemon::UNIX;

     # arguments will be passed to IO::Socket::UNIX, but Listen=>1 and
     # Type=>SOCK_STREAM will be added by default. also, HTTP::Daemon::UNIX will try
     # to delete stale socket first, for convenience.
     my $d = HTTP::Daemon::UNIX->new(Local => "/path/to/unix.sock");

     # will print something like: "http:path/to/unix.sock"
     print "Please contact me at: <URL:", $d->url, ">\n";

     # after that, use like you would use HTTP::Daemon
     while (my $c = $d->accept) {
         while (my $r = $c->get_request) {
             if ($r->method eq 'GET' and $r->uri->path eq "/xyzzy") {
                 # remember, this is *not* recommended practice :-)
                 $c->send_file_response("/etc/passwd");
             } else {
                 $c->send_error(RC_FORBIDDEN);
             }
         }
         $c->close;
         undef($c);
     }

     # client side code, using LWP::Protocol::http::SocketUnixAlt
     use LWP::Protocol::http::SocketUnixAlt;
     use LWP::UserAgent;
     use HTTP::Request::Common;

     my $ua = LWP::UserAgent->new;
     my $orig_imp = LWP::Protocol::implementor("http");
     LWP::Protocol::implementor(http => 'LWP::Protocol::http::SocketUnixAlt');
     my $resp = $ua->request(GET "http:path/to/unix.sock//uri/path");
     LWP::Protocol::implementor(http => $orig_imp);

DESCRIPTION
    This is a quick hack to enable HTTP::Daemon to serve requests over Unix
    sockets, by mixing in IO::Socket::UNIX and HTTP::Daemon as parents to
    HTTP::Daemon::UNIX and overriding IO::Socket::INET-related stuffs.

    Basic stuffs seem to be working, but this module has not been tested
    extensively, so beware that things might blow up in your face.

SEE ALSO
    HTTP::Daemon

    LWP::Protocol::http::SocketUnixAlt

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/HTTP-Daemon-UNIX>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-HTTP-Daemon-UNIX>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=HTTP-Daemon-UNIX>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for HTTP-Daemon-UNIX

0.06    2014-12-05 (PERLANCAR)

	- No functional changes.

	- Use new name of renamed module SHARYANTO::IO::Socket::UNIX::Util
	  -> IO::Socket::UNIX::Util.


0.05    2014-07-06 (SHARYANTO)

        - No functional changes.

        - Extract socket creation routine to SHARYANTO::IO::Socket::UNIX::Util.


0.04    2012-05-23 (SHARYANTO)

        - No functional changes.

        - Make sure connect error messages are in English.


0.03    2011-08-20 (SHARYANTO)

        - Uses IO::Handle::Record for peercred().


0.02    2011-08-04 (SHARYANTO)

        - No functional changes. Add client-side example and mention
          LWP::Protocol::http::SocketUnixAlt.


0.01    2011-07-06 (SHARYANTO)

        - First release.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.025.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "HTTP::Daemon over Unix sockets",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTTP-Daemon-UNIX",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "HTTP::Daemon::UNIX",
  "PREREQ_PM" => {
    "HTTP::Daemon" => 0,
    "IO::Handle::Record" => 0,
    "IO::Socket::UNIX::Util" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Daemon" => 0,
  "IO::Handle" => 0,
  "IO::Handle::Record" => 0,
  "IO::Socket::UNIX::Util" => 0,
  "IPC::Open3" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/HTTP/Daemon/UNIX.pm
package HTTP::Daemon::UNIX;

our $DATE = '2014-12-05'; # DATE
our $VERSION = '0.05'; # VERSION

use 5.010001;
use strict;
use warnings;

use HTTP::Daemon;
use IO::Handle::Record; # for peercred()
use IO::Socket::UNIX::Util qw(create_unix_socket);

our @ISA = qw(HTTP::Daemon IO::Socket::UNIX);

sub new {
    my ($class, %args) = @_;

    # XXX normalize arg case first

    my $sock = SHARYANTO::IO::Socket::UNIX::Util::create_unix_socket(
        $args{Local});
    bless $sock, $class;
}

sub url {
    my ($self) = @_;
    my $hostpath = $self->hostpath;
    $hostpath =~ s!^/!!;
    my $url = $self->_default_scheme . ":" . $hostpath;

    # note: my LWP::Protocol::http::SocketUnixAlt requires this syntax ("//"
    # separates the Unix socket path and URI):
    # http:abs/path/to/unix.sock//uri/path
}

1;
# ABSTRACT: HTTP::Daemon over Unix sockets

__END__

=pod

=encoding UTF-8

=head1 NAME

HTTP::Daemon::UNIX - HTTP::Daemon over Unix sockets

=head1 VERSION

This document describes version 0.05 of HTTP::Daemon::UNIX (from Perl distribution HTTP-Daemon-UNIX), released on 2014-12-05.

=head1 SYNOPSIS

 use HTTP::Daemon::UNIX;

 # arguments will be passed to IO::Socket::UNIX, but Listen=>1 and
 # Type=>SOCK_STREAM will be added by default. also, HTTP::Daemon::UNIX will try
 # to delete stale socket first, for convenience.
 my $d = HTTP::Daemon::UNIX->new(Local => "/path/to/unix.sock");

 # will print something like: "http:path/to/unix.sock"
 print "Please contact me at: <URL:", $d->url, ">\n";

 # after that, use like you would use HTTP::Daemon
 while (my $c = $d->accept) {
     while (my $r = $c->get_request) {
         if ($r->method eq 'GET' and $r->uri->path eq "/xyzzy") {
             # remember, this is *not* recommended practice :-)
             $c->send_file_response("/etc/passwd");
         } else {
             $c->send_error(RC_FORBIDDEN);
         }
     }
     $c->close;
     undef($c);
 }

 # client side code, using LWP::Protocol::http::SocketUnixAlt
 use LWP::Protocol::http::SocketUnixAlt;
 use LWP::UserAgent;
 use HTTP::Request::Common;

 my $ua = LWP::UserAgent->new;
 my $orig_imp = LWP::Protocol::implementor("http");
 LWP::Protocol::implementor(http => 'LWP::Protocol::http::SocketUnixAlt');
 my $resp = $ua->request(GET "http:path/to/unix.sock//uri/path");
 LWP::Protocol::implementor(http => $orig_imp);

=head1 DESCRIPTION

This is a quick hack to enable L<HTTP::Daemon> to serve requests over Unix
sockets, by mixing in L<IO::Socket::UNIX> and HTTP::Daemon as parents to
L<HTTP::Daemon::UNIX> and overriding IO::Socket::INET-related stuffs.

Basic stuffs seem to be working, but this module has not been tested
extensively, so beware that things might blow up in your face.

=head1 SEE ALSO

L<HTTP::Daemon>

L<LWP::Protocol::http::SocketUnixAlt>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/HTTP-Daemon-UNIX>.

=head1 SOURCE

Source repository is at L<https://github.com/perlancar/perl-HTTP-Daemon-UNIX>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=HTTP-Daemon-UNIX>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

perlancar <perlancar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by perlancar@cpan.org.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
